### PR TITLE
[jest] add `jest.replaceProperty` method types

### PR DESCRIPTION
**Pull request by rpvsilva**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://jestjs.io/docs/jest-object#jestreplacepropertyobject-propertykey-value
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

**Comment by mrazauskas** [comment]
Thanks for coming back to this! I had no luck with this feature, but perhaps you could reuse tests from this commit https://github.com/DefinitelyTyped/DefinitelyTyped/pull/64052/commits/75bbfbd011749c7d877b09c30951aa54fbad06a0?

File: types/jest/index.d.ts [hunk 1]
```diff
@@ -1,4 +1,4 @@
-// Type definitions for Jest 29.4
+// Type definitions for Jest 29.5
```
Comment (mrazauskas) [inline types/jest/index.d.ts:1]: Why version bump? This is feature was added in [Jest v29.4](https://github.com/facebook/jest/releases/tag/v29.4.0).

```suggestion
// Type definitions for Jest 29.4
```

**Comment by DangerBotOSS** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## gestalt ([<kbd>unpkg</kbd>](https://unpkg.com/browse/gestalt@latest/))
was missing the following properties:
1. Fragment
1. jsx
1. jsxs
1. Collection
1. ExperimentProvider

<details>
<summary>as well as these 3 other properties...</summary>
<p>IconButtonFloating, ScrollFetch, useColorScheme</p>
</details>

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 0a20fb981c4d695ad16cdc0345c38a820fd9a851
</p>

File: types/jest/index.d.ts (hunk 1 above)
Comment (rpvsilva) [reply types/jest/index.d.ts]: Reverted this change, thanks 😉

**Comment by rpvsilva** [comment]
> Thanks for coming back to this! I had no luck with this feature, but perhaps you could reuse tests from this commit [75bbfbd](https://github.com/DefinitelyTyped/DefinitelyTyped/commit/75bbfbd011749c7d877b09c30951aa54fbad06a0)?

Just added your tests and it's working

**Comment by mrazauskas** [comment]
Nice. The test below is missing:

```ts
// @ts-expect-error: method cannot be replaced
jest.replaceProperty(replaceObjectA, 'method', () => {});
```

Of course, I understand the complexity behind it, but still... if Jest's builtin types (i.e. `import {jest} from '@jest/globals';`) can do this, why `@types/jest` cannot?

**Comment by rpvsilva** [comment]
Why can't you replace a method type? Is there any reason?

**Comment by mrazauskas** [comment]
Good point! That is even documented as a use case (see last line):

Screenshot 2023-03-14 at 20 16 22

These tests were copied and adapted these tests from Jest repo (I will fix it there): 

https://github.com/facebook/jest/blob/94c06ef0aa9b327f3c400610b861e7308b29ee0d/packages/jest-mock/__typetests__/mock-functions.test.ts#L511

File: types/jest/index.d.ts [hunk 2]
```diff
@@ -286,6 +286,11 @@ declare namespace jest {
      * This only works with jest-circus!
      */
     function retryTimes(numRetries: number, options?: { logErrorsBeforeRetry?: boolean }): typeof jest;
+    /**
+     * Replace object[propertyKey] with a value.
+     * The property must already exist on the object. The same property might be replaced multiple times. Returns a Jest replaced property.
+     */
+    function replaceProperty<T, K extends keyof T>(obj: T, key: K, value: T[K]): ReplaceProperty<T[K]>;
```
Comment (mrazauskas) [inline types/jest/index.d.ts:293]: Could you copy JSDoc for `replaceProperty` and `resetAllMocks` from the same commit: https://github.com/DefinitelyTyped/DefinitelyTyped/commit/75bbfbd011749c7d877b09c30951aa54fbad06a0

I took them from Jest repo.

**Review by mrazauskas — CHANGES_REQUESTED** [review]
Could you fix JSDocs, please? The rest looks good to me.

File: types/jest/index.d.ts (hunk 2 above)
Comment (rpvsilva) [reply types/jest/index.d.ts:295]: Done

Comment (mrazauskas) [reply types/jest/index.d.ts:295]: Also the comment above `resetAllMocks`, please. It must mention `replaceProperty`.

**Review by mrazauskas — CHANGES_REQUESTED** [review]
Could you add this (from the same https://github.com/DefinitelyTyped/DefinitelyTyped/commit/75bbfbd011749c7d877b09c30951aa54fbad06a0) please:

Screenshot 2023-03-14 at 21 00 21

It is important to mention, because `jest.restoreAllMocks` also restores replaced properties.

**Review by mrazauskas — APPROVED** [review]
Wonderful! Thank you.

**Comment by GerkinDev** [comment]
> Why can't you replace a method type? Is there any reason? Because if you actually try it using jest it works, since we are only typing the function and not limiting the functionality I don't know if it makes sense to restrict it

Just thinking: I would also prefer suggesting to the user to use `spyOn` instead to limit user errors; one of the main reasons for using TS is to help users. Maybe this can be done via an overload signature. But I'm OK if not, the comment is enough

**Comment by mrazauskas** [comment]
What you mean by user errors?

It might be that the name of this method will change at some point. There are ideas to call it `jest.setProperty()` or `jest.stubProperty()`. Actually the name of `jest.replaceProperty()` also makes it clear to me that this method does not spy, simply stubs a property.

Both `jest.spyOn()` and `jest.replaceProperty()` can be used on objects. The result is different, but it is clear from names. Or I missed something?

**Review by jakebailey — APPROVED** [review]

**Comment by rpvsilva** [comment]
Ready to merge

**Pull request merged**